packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt

```diff
@@ -221,7 +214,26 @@ internal object RealmObjectHelper {
         value: R
     ) {
         val o = obj.`$realm$ObjectPointer` ?: throw IllegalStateException("Invalid/deleted object")
-        RealmInterop.realm_set_value(o, key, value, false)
+        try {
+            // TODO Consider making a RealmValue cinterop type and move the various to_realm_value
+            //  implementations in the various platform RealmInterops here to eliminate
+            //  RealmObjectInterop and make cinterop operate on primitive values and native pointers
+            //  only. This relates to the overall concern of having a generic path for getter/setter
+            //  instead of generating a typed path for each type.
+            RealmInterop.realm_set_value(o, key, value, false)
+            // The catch block should catch specific Core exceptions and rethrow them as Kotlin exceptions.
+            // Core exceptions meaning might differ depending on the context, by rethrowing we can add some context related
+            // info that might help users to understand the exception.
+        } catch (exception: RealmCorePropertyNotNullableException) {
+            throw IllegalArgumentException("Required property `${obj.`$realm$ClassName`}.${obj.`$realm$metadata`!!.get(key)!!.name}` cannot be null")
+        } catch (exception: RealmCorePropertyTypeMismatchException) {
+            throw IllegalArgumentException("Property `${obj.`$realm$ClassName`}.${obj.`$realm$metadata`!!.get(key)!!.name}` cannot be assigned with value '$value' of wrong type")
+        } catch (exception: RealmCoreException) {
+            throw IllegalStateException(
```

Comment:
Are we sure this is the only exception possible here? Or should we match on the Core message? (ugly and error-prone).

Reply:
Again just moved these. But I can try to improve it if we need.

Reply:
If they where just moved, we can probably just cycle back to this as part of future improvements 👍